File: VB/WpfApplication1/CustomBarElements.vb
Imports DevExpress.Xpf.Bars
Imports System
Imports System.Windows.Controls
Imports System.Windows.Media

Namespace WpfApplication1

    Public Module CustomBarElementsRegistrator

        Public Sub Register()
            Call BarItemLinkCreator.Default.RegisterObject(GetType(MyBarButtonItem), GetType(MyBarButtonItemLink), Function(link) New MyBarButtonItemLink())
            Call BarItemLinkControlCreator.Default.RegisterObject(GetType(MyBarButtonItemLink), GetType(MyBarButtonItemLinkControl), Function(link) New MyBarButtonItemLinkControl(CType(link, MyBarButtonItemLink)))
        End Sub
    End Module

    Public Class MyBarButtonItemLinkControl
        Inherits BarButtonItemLinkControl

        Public Sub New()
        End Sub

        Public Sub New(ByVal link As MyBarButtonItemLink)
            MyBase.New(link)
        End Sub

        Public Overrides Sub OnApplyTemplate()
            MyBase.OnApplyTemplate()
            'demo test
            Me.LayoutPanel.Background = Brushes.Red
        End Sub
    End Class

    Public Class MyBarButtonItemLink
        Inherits BarButtonItemLink

    End Class

    Public Class MyBarButtonItem
        Inherits BarButtonItem

        Protected Overrides Function GetLinkType() As Type
            Return GetType(MyBarButtonItemLink)
        End Function
    End Class
End Namespace

File: CS/WpfApplication1/CustomBarElements.cs
using DevExpress.Xpf.Bars;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using System.Windows;
using System.Windows.Controls;
using System.Windows.Media;

namespace WpfApplication1
{
    public static class CustomBarElementsRegistrator
    {
        public static void Register()
        {
            BarItemLinkCreator.Default.RegisterObject(typeof(MyBarButtonItem), typeof(MyBarButtonItemLink), link => new MyBarButtonItemLink());
            BarItemLinkControlCreator.Default.RegisterObject(typeof(MyBarButtonItemLink), typeof(MyBarButtonItemLinkControl), link => new MyBarButtonItemLinkControl((MyBarButtonItemLink)link));
        }
    }

    public class MyBarButtonItemLinkControl : BarButtonItemLinkControl
    {
        public MyBarButtonItemLinkControl()
        {
        }
        public MyBarButtonItemLinkControl(MyBarButtonItemLink link)
            : base(link)
        {
            
        }

        public override void OnApplyTemplate()
        {
            base.OnApplyTemplate();

            //demo test
            this.LayoutPanel.Background = Brushes.Red;
        }
    }

    public class MyBarButtonItemLink : BarButtonItemLink
    {

    }


    public class MyBarButtonItem : BarButtonItem
    {
        protected override Type GetLinkType()
        {
            return typeof(MyBarButtonItemLink);
        }
    }

}
